backend/app/routers/authorization.py

```diff
@@ -18,23 +23,62 @@ async def save_authorization(
     user=Depends(keycloak_auth.get_current_username),
     db: MongoClient = Depends(dependencies.get_db),
 ):
-    """Save authorization info in Mongo. This is a triple of dataset_id/user_id/role."""
-    authorization_db = AuthorizationDB(**authorization_in.dict(), creator=user)
+    """Save authorization info in Mongo. This is a triple of dataset_id/user_id/role/group_id."""
+
+    # Check all connection and abort if any one of them is not available
+    if db is None:
+        raise HTTPException(status_code=503, detail="Service not available")
+        return
+
+    # Retrieve users from groups in mongo
+    user_ids: List[EmailStr] = []
+    group_q_list = db["groups"].find({"_id": {"$in": authorization_in.group_ids}})
+    if group_q_list is not None:
+        async for group_q in group_q_list:
+            group = GroupOut.from_mongo(group_q)
+            user_ids = list(map(lambda user: user.user.email, group.userList))
+    else:
+        raise HTTPException(
+            status_code=404, detail=f"Group {authorization_in.group_ids} not found"
+        )
+
+    authorization_dict = authorization_in.dict()
+    authorization_dict["user_ids"] = list(set(user_ids))
+    authorization_db = AuthorizationDB(**authorization_dict, creator=user)
     new_authorization = await db["authorization"].insert_one(
         authorization_db.to_mongo()
     )
     found = await db["authorization"].find_one({"_id": new_authorization.inserted_id})
     return AuthorizationDB.from_mongo(found)
 
 
-@router.get("/datasets/{dataset_id}/role", response_model=AuthorizationBase)
+@router.get("/datasets/{dataset_id}/role", response_model=AuthorizationDB)
 async def get_dataset_role(
     dataset_id: str,
     current_user=Depends(get_current_username),
-    role: RoleType = Depends(get_role),
+    db: MongoClient = Depends(get_db),
 ):
     """Retrieve role of user for a specific dataset."""
-    return AuthorizationBase(dataset_id=dataset_id, user_id=current_user, role=role)
+
+    # Get group id and the associated userList from authorization
+    if (
+        authorization_q := await db["authorization"].find_one(
+            {"dataset_id": ObjectId(dataset_id)}
+        )
+    ) is not None:
+        authorization = AuthorizationDB.from_mongo(authorization_q)
+
+        if current_user in authorization.user_ids:
```

Comment:
Here we could check the groups, by retrieving the list of groups a user is in and checking if any of those groups are in the Authorization group list.

Reply:
That works too but I am wondering if the way I have implemented it now will not work.